### PR TITLE
Fix serial device on ssh-root console in select_serial_terminal

### PR DESCRIPTION
**Pull request by dzedro**
- Related ticket: https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/8152
- Fail: http://openqa.qam.suse.cz/tests/2706#step/boot_ltp/5
- Verification run:
http://openqa.qam.suse.cz/tests/2765

### CODE REVIEW

**Comment by pevik** [comment]
@dzedro Well, others moved use of `use_ssh_serial_console` into `activate_console()`. Originally it was added in  849cb56c2e380b89bc3cabd35a3ccbca653d530b, later fixes for missing code are e.g. in c823e3d6cab90ecaa0e60c1ef351a4b5a00c981c. Please add your fix there.

**Comment by pevik** [comment]
And if it's not possible to add another console to https://github.com/os-autoinst/os-autoinst-distri-opensuse/blob/master/lib/susedistribution.pm#L668, which would fix issue, it'd be better to use `use_ssh_serial_console` below `$console = 'root-ssh';` (DRY)

**Comment by pevik** [comment]
@okurz you're behind original idea of using `use_ssh_serial_console`, please take care of it.

**Comment by pevik** [comment]
I guess this effort originated in https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/5239
https://progress.opensuse.org/issues/38423.

Not sure, what is the needed change for this particular case.

**Comment by pevik** [comment]
@dzedro any progress here? IMHO @czerw has some problems with `spvm`.
@czerw FYI

**Comment by czerw** [comment]
> @dzedro any progress here? IMHO @czerw has some problems with `spvm`.

@pevik Thanks for notice about this PR, but it doesn't look similar.

**Comment by dzedro** [comment]
no progress now, I used old function to make it work on our instance, and what I tried with activate_console didn't work, it should work because the code was executed but the serialdev was still ttyS1

**Comment by pevik** [comment]
@dzedro looks like the original version was really the only solution. I'm sorry for refusing it.

**Comment by pevik** [comment]
@dzedro I'd consider extracting following code into separate function. It's used already on 2 places, you're going to add it to yet another :).

```
 $serialdev = 'sshserial';
set_var('SERIALDEV', $serialdev);
bmwqemu::save_vars();
```

**Comment by pevik** [comment]
@dzedro @okurz BTW, looking at `64bit-mlx` at https://openqa.suse.de/admin/machines, these have `SERIALDEV=sshserial` set. Wouldn't be better just fix all IPMI machines instead of setting it with `use_ssh_serial_console`? The only relevant part from it is `select_console('root-ssh') (which is also done in `select_serial_terminal()`).
@frankenmichl FYI, can you comment please?

**Comment by frankenmichl** [comment]
Yes, we should set this in the machine definition, as these are static. And imho setting such variables makes debugging issues hard, as the variables get overwritten and you don't notice it.
I would prefer not setting SERIALDEV from test code.

File: lib/susedistribution.pm [hunk 1]
```diff
@@ -742,6 +742,9 @@ sub activate_console {
         ensure_user($user);
         assert_screen(["text-logged-in-$user", "text-login"], 60);
         $self->set_standard_prompt($user, skip_set_standard_prompt => $args{skip_set_standard_prompt});
+        $serialdev = 'sshserial';
+        set_var('SERIALDEV', $serialdev);
```
Comment (jlausuch) [inline lib/susedistribution.pm]:
```suggestion
        set_var('SERIALDEV', 'sshserial');
```

and remove line above

Comment (pevik) [reply lib/susedistribution.pm]: Yep. But I'd be really for fixing machines' setup and get rid of `use_ssh_serial_console`.

**Comment by dzedro** [comment]
@pevik ipmi is not qemu and installation tests which haveto use sol will not work with sshserial!

File: lib/susedistribution.pm (hunk 1 above)
Comment (dzedro) [reply lib/susedistribution.pm]: This will not work, you have to change global variable $sshseriel, I started with your suggestion

Comment (pevik) [reply lib/susedistribution.pm]: Having to set a global variable while there is openQA environment variable, huh :). I just love this mess :).

**Comment by pevik** [comment]
> @pevik ipmi is not qemu and installation tests which have to use sol will not work with sshserial!

Not sure if I understand you. My suggestion was to set `SERIALDEV` properly to each machine. Or is it that some machine is used in more backends therefore (or for any other reason) it needs to use different `SERIALDEV`?

**Comment by dzedro** [comment]
ipmi machines have mostly SERIALDEV=ttyS1 like here http://openqa.qam.suse.cz/tests/2760#step/boot_from_pxe/5 when you want to install the machine, when you just want to run tests on already installed ipmi machine, you can use SERIALDEV=sshserial

File: lib/susedistribution.pm (hunk 1 above)
Comment (dzedro) [reply lib/susedistribution.pm]: not set, change ...

**Comment by pevik** [comment]
> ipmi machines have mostly SERIALDEV=ttyS1 like here http://openqa.qam.suse.cz/tests/2760#step/boot_from_pxe/5 when you want to install the machine, when you just want to run tests on already installed ipmi machine, you can use SERIALDEV=sshserial

@frankenmichl that explains why somebody set sonic & tails `SERIALDEV=ttyS1`.

@dzedro Interesting, that confirm my assumption that there is something different for install and post install mode (I expected `root-ssh` vs. `root-console`). That shows that setting `SERIALDEV` in machine is not a good idea (we use it for sonic & tails, but we use only installation with iPXE, therefore `SERIALDEV=sshserial` is correct for our tests).

Is there any openQA environment variable (or anything else), which could help to decide `SERIALDEV` automatically?

**Comment by dzedro** [comment]
> > ipmi machines have mostly SERIALDEV=ttyS1 like here http://openqa.qam.suse.cz/tests/2760#step/boot_from_pxe/5 when you want to install the machine, when you just want to run tests on already installed ipmi machine, you can use SERIALDEV=sshserial
> 
> @frankenmichl that explains why somebody set sonic & tails `SERIALDEV=ttyS1`.
> 
> @dzedro Interesting, that confirm my assumption that there is something different for install and post install mode (I expected `root-ssh` vs. `root-console`). That shows that setting `SERIALDEV` in machine is not a good idea (we use it for sonic & tails, but we use only installation with iPXE, therefore `SERIALDEV=sshserial` is correct for our tests).
> 
> Is there any openQA environment variable (or anything else), which could help to decide `SERIALDEV` automatically?

Yes if you ca somehow distinguish tests which do installation of ipmi machine otherwise no ?

**Comment by frankenmichl** [comment]
The machines not using the IPMI console always also have the variable IPXE defined to '1'. But maybe we can come up with something smarter here?

**Comment by ldevulder** [comment]
I tested this PR (temporary added in PR #8253) and I can confirm that it works well: https://openqa.suse.de/tests/3343083

Without it I had issue with console device (`/dev/ttyS1` instead of `/dev/sshserial`: https://openqa.suse.de/tests/3330196#step/wizard_hana_install/1

**Comment by pevik** [comment]
Let's merge it.

**Comment by pevik** [comment]
Removed some duplicity https://github.com/os-autoinst/os-autoinst-distri-opensuse/pull/8424

**Pull request merged**